sphinx/screens-detail/payments/payment-template/payment-template/src/main/java/chat/sphinx/payment_template/ui/adapter/PaymentTemplateAdapter.kt

```diff
@@ -0,0 +1,116 @@
+import chat.sphinx.wrapper_message_media.token.MediaHost
+import io.matthewnelson.android_feature_screens.util.gone
+import io.matthewnelson.android_feature_screens.util.visible
+import io.matthewnelson.android_feature_viewmodel.util.OnStopSupervisor
+import kotlinx.coroutines.launch
+import kotlin.collections.ArrayList
+
+internal class PaymentTemplateAdapter(
+    private val recyclerView: RecyclerView,
+    private val imageLoader: ImageLoader<ImageView>,
+    private val onStopSupervisor: OnStopSupervisor,
+    private val viewModel: PaymentTemplateViewModel,
+): RecyclerView.Adapter<PaymentTemplateAdapter.PaymentTemplateViewHolder>(), DefaultLifecycleObserver {
+
+    companion object {
+        const val PADDING_ITEMS_COUNT: Int = 2
+        const val NO_TEMPLATE_ITEMS_COUNT: Int = 1
+    }
+
+    private var paymentTemplates = ArrayList<PaymentTemplate>()
+
+    fun setItems(newItems: List<PaymentTemplate>) {
+        paymentTemplates.addAll(newItems)
+        notifyDataSetChanged()
+    }
+
+    override fun getItemCount(): Int {
+        return paymentTemplates.size + PADDING_ITEMS_COUNT + NO_TEMPLATE_ITEMS_COUNT
+    }
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PaymentTemplateAdapter.PaymentTemplateViewHolder {
+        val binding = LayoutPaymentTemplateHolderBinding.inflate(
+            LayoutInflater.from(parent.context),
+            parent,
+            false
+        )
+
+        return PaymentTemplateViewHolder(binding)
+    }
+
+    override fun onBindViewHolder(holder: PaymentTemplateAdapter.PaymentTemplateViewHolder, position: Int) {
+        holder.bind(position)
+    }
+
+    inner class PaymentTemplateViewHolder(
+        private val binding: LayoutPaymentTemplateHolderBinding
+    ): RecyclerView.ViewHolder(binding.root), DefaultLifecycleObserver {
+
+        fun bind(position: Int) {
+            binding.apply {
+
+                val itemWidth = recyclerView.context.resources.getDimension(R.dimen.payment_template_recycler_view_item_width).toInt()
+                val paddingItemWidth = ((recyclerView.width - itemWidth) / 2)
+
+                val isStartPaddingItem = position == 0
+                val isEndPaddingItem = position == paymentTemplates.size + 2
+
+                if (isStartPaddingItem || isEndPaddingItem) {
+                    root.updateLayoutParams { width = paddingItemWidth }
+
+                    imageViewTemplate.gone
+                } else {
+                    root.updateLayoutParams { width = itemWidth }
+
+                    imageViewTemplate.visible
+
+                    paymentTemplates.getOrNull(position - 2)?.let { paymentTemplate ->
+                        paymentTemplate.getTemplateUrl(MediaHost.DEFAULT.value)?.let { url ->
+
+                            val token = AuthenticationToken(paymentTemplate.token)
+
+                            onStopSupervisor.scope.launch(viewModel.dispatchers.mainImmediate) {
```

Comment:
improper use of `onStopSupervisor`. remove. Use lifecycle scope, track job + disposable to cancle upon holder re-bind.